[25.12] treewide: APK provides-related backports

GeorgeSapkin opened the pull request:
Backport #20819 and #21265 to 25.12.

Depends on:
- https://github.com/openwrt/openwrt/pull/21265
- https://github.com/openwrt/packages/pull/28155
- https://github.com/openwrt/packages/pull/28145
- https://github.com/openwrt/packages/pull/28158

Related:
- https://github.com/openwrt/openwrt/issues/21257

GeorgeSapkin committed:
build: fix apk packaging and ABI-versioning

The updated logic for the APK dependencies and provides is as follows:

- If ABI version is defined:
  - package is named `package_name-ABI_version`
  - package implicitly provides
    `package_name-ABI_version=package_version`
    this implies that only one version of a package per ABI can be
    installed at the same time
  - additionally provide `package_name` so multiple packages can be
    looked up by its base name
  - for each `provides`, provide `provide-ABI_version=package_version`
    this implies that only one version of a provide can be installed at
    the same time

- else if ABI version is _not_ defined
  - package is named `package_name`
  - package implicitly provides `package_name=package_version`
    this implies that only one version of a package can be installed at
    the same time
  - if `alternatives` is defined
    - for each `provides`, provide `provide`
      this implies that multiple versions of a provide can be installed
      at the same time
  - else if `alternatives` is _not_ defined
    - for each `provides`, provide `provide=package_version`
      this implies that only one version of a provide can be installed
      at the same time

Both cases a package can be looked up by its base name.

ABI version `alternatives`, `conffiles`, `conffiles_static`, `list` and
`rusers` files so multiple versions of the same ABI package can be
installed side by side, and so they don't overwrite each other's
packaging files.

ABI version `EXTRA_DEPENDS` so dependencies can be correctly looked up
using the existing OpenWrt semantics without the ABI specified. This is
needed since ABI-versioned libraries no longer provide
`package_name=package_version`, so that they can be installed side by
side.

Remove duplicate dependencies when `EXTRA_DEPENDS` specifies a versioned
one that is already in `DEPENDS`.

ABI is defined
------------------------------------------------------------------------

`libsqlite3` has `PROVIDES` set to `libfake` and has two different ABI
versions installed. `libfake` is just an example to demonstrate the
mechanics, as the library can already be depended upon using e.g.
`libsqlite3-0=3.51.0-r1`. Note the ABI-versioned lists.

```
root@OpenWrt:/tmp# apk add --allow-untrusted ./libsqlite3-0-3.51.0-r1.apk
(1/1) Installing libsqlite3-0 (3.51.0-r1)
libsqlite3-0-3.51.0-r1.post-install: Executing script...
OK: 22 MiB in 157 packages

root@OpenWrt:/tmp# apk add --allow-untrusted ./libsqlite3-1-4.00.0-r1.apk
(1/1) Installing libsqlite3-1 (4.00.0-r1)
libsqlite3-1-4.00.0-r1.post-install: Executing script...
OK: 23 MiB in 158 packages

root@OpenWrt:/tmp# apk query --fields name,version,contents,provides libsqlite3-0 libsqlite3-1
Name: libsqlite3-0
Version: 3.51.0-r2
Provides: libfake-0=3.51.0-r2 libsqlite3
Contents:
  lib/apk/packages/libsqlite3-0.list
  usr/lib/libsqlite3.so.0
  usr/lib/libsqlite3.so.3.51.0

Name: libsqlite3-1
Version: 4.00.0-r1
Provides: libfake-1=4.00.0-r1 libsqlite3
Contents:
  lib/apk/packages/libsqlite3-1.list
  usr/lib/libsqlite3.so.1
  usr/lib/libsqlite3.so.4.00.0

root@OpenWrt:/tmp# ls -lh /usr/lib/libsqlite3.so.*
lrwxrwxrwx    1 root     root          20 Nov 20 00:23 /usr/lib/libsqlite3.so.0 -> libsqlite3.so.3.51.0
lrwxrwxrwx    1 root     root          20 Nov 20 00:27 /usr/lib/libsqlite3.so.1 -> libsqlite3.so.4.00.0
-rwxr-xr-x    1 root     root        1.0M Nov  6 18:19 /usr/lib/libsqlite3.so.3.51.0
-rwxr-xr-x    1 root     root        1.0M Nov  6 18:19 /usr/lib/libsqlite3.so.4.00.0
```

ABI is not defined
------------------------------------------------------------------------

Both `avahi-dbus-daemon` and `avahi-nodbus-daemon` provide `avahi-daemon`,
but have no ABI specified. This results in `avahi-daemon=0.8-r11` provides
for both packages and only one being able to be installed at the same time:

```
root@OpenWrt:/tmp# apk add --allow-untrusted ./avahi-nodbus-daemon-0.8-r11.apk
(1/4) Installing libavahi-nodbus-support (0.8-r10)
libavahi-nodbus-support-0.8-r10.post-install: Executing script...
(2/4) Installing libdaemon (0.14-r5)
libdaemon-0.14-r5.post-install: Executing script...
(3/4) Installing libexpat (2.7.3-r1)
libexpat-2.7.3-r1.post-install: Executing script...
(4/4) Installing avahi-nodbus-daemon (0.8-r11)
avahi-nodbus-daemon-0.8-r11.post-install: Executing script...
23 MiB in 160 packages

root@OpenWrt:/tmp# apk query --fields provides avahi-nodbus-daemon
Provides: avahi-daemon=0.8-r11

root@OpenWrt:/tmp# apk add --allow-untrusted ./avahi-dbus-daemon-0.8-r11.apk
ERROR: unable to select packages:
  avahi-dbus-daemon-0.8-r11:
    conflicts: avahi-nodbus-daemon-0.8-r11[avahi-daemon=0.8-r11]
    satisfies: world[avahi-dbus-daemon><Q1R111s+ke9Vf+eCxDHX2BZVUK54Q=]
  avahi-nodbus-daemon-0.8-r11:
    conflicts: avahi-dbus-daemon-0.8-r11[avahi-daemon=0.8-r11]
    satisfies: world[avahi-nodbus-daemon><Q1BAu7nLI2MgRabpveLTGO2ksQz7E=]
```

Provides and alternatives
------------------------------------------------------------------------

Both `uclient-fetch` and `wget-nossl` provide `wget` and specify
alternatives, so provides are not versioned and both packages can be
installed at the same time:

```
root@OpenWrt:/tmp# apk query --fields name,version,contents,provides uclient-fetch wget-nossl
Name: uclient-fetch
Version: 2025.10.03~dc909ca7-r1
Provides: wget
Contents:
  bin/uclient-fetch
  lib/apk/packages/uclient-fetch.alternatives
  lib/apk/packages/uclient-fetch.list

Name: wget-nossl
Version: 1.25.0-r1
Provides: gnu-wget wget
Contents:
  lib/apk/packages/wget-nossl.alternatives
  lib/apk/packages/wget-nossl.list
  usr/libexec/wget-nossl

```

Fixes: https://github.com/openwrt/openwrt/issues/20582
Fixes: https://github.com/openwrt/openwrt/issues/20802
Signed-off-by: George Sapkin <george@sapk.in>
Link: https://github.com/openwrt/openwrt/pull/20819
(cherry picked from commit 18029977f65e11bafaad501399ad42a66d3baa10)
Link: https://github.com/openwrt/openwrt/pull/21253
Signed-off-by: Robert Marko <robimarko@gmail.com>

GeorgeSapkin commented:
There are a few packages, e.g. nginx, that have _creative_ use of `PROVIDES`, I'm looking into it and fixing what doesn't make sense, but realistically its not feasible for me to build and install every combination of packages to test this. So the plan is to address issues as they are being reported.

efahl commented:
> its not feasible for me to build and install every combination of packages to test this.

Wait, why not?  😂  

There's recent precedent for this: when apk changed package version format requirements, it was mostly left to the package maintainers to clean up.  I'd say this is pretty much the same situation...

GeorgeSapkin commented:
> Wait, why not? 😂

TBH, this would be cool project to make. That is to check the dependency graph without building, I'm trying not think about it too much, so I don't pick it up.

The challenge with this PR is that the apk fix really should go into 25.12, but there are some breakages here and there, like nginx. And I fear there will be incompatible configuration that can't be upgraded to 25.12 without manual intervention as you mention. There are already multiple packages that had version changes to match upstream versioning schemes, like sqlite. Might be something to keep in mind for owut, if it doesn't do that already.

GeorgeSapkin committed:
build: fix missing PKG_INFO_DIR

If the first built package has an ABI, PKG_INFO_DIR might not exist, so
ensure it does.

Signed-off-by: George Sapkin <george@sapk.in>
Link: https://github.com/openwrt/openwrt/pull/21265
(cherry picked from commit 56bc4a0d894f18933c6de2b65a042d4f34f50cd9)
Link: https://github.com/openwrt/openwrt/pull/21253
Signed-off-by: Robert Marko <robimarko@gmail.com>

GeorgeSapkin committed:
toolchain: fix libgcc ABI for libc

libc depends on knowing libgcc's ABI, so it needs to be evaluated first.
Otherwise libc will depend on an ABI-less libgcc causing the rest of the
toolchain to fail.

Building package index...
ERROR: unable to select packages:
  libgcc (virtual):
    note: please select one of the 'provided by'
          packages explicitly
    provided by: libgcc1
    required by: world[libgcc]

Before:

libc fused dependencies: libgcc
libgcc fused dependencies:
libatomic fused dependencies: libgcc1

After:

libgcc fused dependencies:
libc fused dependencies: libgcc1
libatomic fused dependencies: libgcc1

Signed-off-by: George Sapkin <george@sapk.in>
Link: https://github.com/openwrt/openwrt/pull/21265
(cherry picked from commit 5e32bfbf4bbe8fa4517f27d91d5f53177f7b836a)
Link: https://github.com/openwrt/openwrt/pull/21253
Signed-off-by: Robert Marko <robimarko@gmail.com>

teleostnacl commented:
I have applied these changes, but they appear to have no effect on this issue: https://github.com/openwrt/openwrt/issues/21257

GeorgeSapkin commented:
> I have applied these changes, but they appear to have no effect on this issue: #21257

This PR has no fix for that issue. The related part is to keep track of what still need to be looked at.

teleostnacl commented:
> > I have applied these changes, but they appear to have no effect on this issue: #21257
> 
> This PR has no fix for that issue. The related part is to keep track of what still need to be looked at.

OK, I got it

bluewavenet commented:
@GeorgeSapkin @robimarko 
It looks like it is now impossible to install any package variants on master/snapshot and 25.12.[anything].
eg:
 - ath10k non-ct drivers
 - iptables-nft
 - dnsmasq-full
 - wpad full or mesh versions
 - libmicrohttp-ssl
 - luci-ssl
 - almost everything with a cut down and a -full variant
 - probably many many more packages

GeorgeSapkin commented:
@bluewavenet with this patchset?

Edit: wrong PR, I meant this one:
- https://github.com/openwrt/openwrt/pull/21288

GeorgeSapkin commented:
@bluewavenet can you share your findings in the following issue? Thanks.
- https://github.com/openwrt/openwrt/issues/21257

systemcrash commented:
I suspect this PR broke uspot e0ea3d73d2cf4d72af107f1419aba3ce287c0080

ping also @BKPepe 

```
[5/6] Building C object CMakeFiles/uspot-das.dir/radius-das.c.o
[6/6] Linking C executable uspot-das
ninja: Entering directory `/builder/shared-workdir/build/sdk/build_dir/target-aarch64_cortex-a53_musl/uspot-2025.11.20~389e48df'
ninja: warning: Ignoring jobserver: Pipe-based protocol is not supported! [--jobserver-auth=3,4]
[0/1] Install the project...
-- Install configuration: "Release"
-- Installing: /builder/shared-workdir/build/sdk/build_dir/target-aarch64_cortex-a53_musl/uspot-2025.11.20~389e48df/ipkg-install/usr/sbin/radius-client
-- Installing: /builder/shared-workdir/build/sdk/build_dir/target-aarch64_cortex-a53_musl/uspot-2025.11.20~389e48df/ipkg-install/usr/sbin/uspot-das
-- Installing: /builder/shared-workdir/build/sdk/build_dir/target-aarch64_cortex-a53_musl/uspot-2025.11.20~389e48df/ipkg-install/usr/lib/libuam.so
ERROR: info field 'depends' has invalid value: dependency format is invalid
ERROR: failed to create package: dependency format is invalid
make[3]: *** [Makefile:136: /builder/shared-workdir/build/sdk/bin/packages/aarch64_cortex-a53/packages/uspotfilter-2025.11.20~389e48df-r1.apk] Error 99
time: package/feeds/packages/uspot/compile#1.80#1.01#5.50
```

The problem(?) being that uspot Makefile has things like:
```Makefile
...
EXTRA_DEPENDS:=ucode (>= 2023.11.07)
...
```

BKPepe commented:
I think the extra space between = and version could be wrong. I will create PR to address it.
